java-checks/src/main/java/org/sonar/java/checks/RedundantThrowsDeclarationCheck.java

```diff
@@ -43,83 +51,128 @@
-          }
-        }
-      }
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.THROW_STATEMENT)) {
+      registerThrownType(((ThrowStatementTree) tree).expression().symbolType());
+    } else if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      Symbol methodSymbol = ((MethodInvocationTree) tree).symbol();
+      processMethodCall(methodSymbol);
+    } else if (tree.is(Tree.Kind.NEW_CLASS)) {
+      Symbol constructorSymbol = ((NewClassTree) tree).constructorSymbol();
+      processMethodCall(constructorSymbol);
+    } else {
+      thrownExceptions.push(new HashSet<Type>());
     }
   }
 
-  private static String getIssueMessage(List<AsmClass> thrownClasses, AsmClass thrownClass) {
-    String thrownClassName = thrownClass.getDisplayName();
-    if (isSubClassOfAny(thrownClass, thrownClasses)) {
-      return "Remove the declaration of thrown exception '" + thrownClassName + "' which is a subclass of another one.";
-    } else if (isSubClassOfRuntimeException(thrownClass)) {
-      return "Remove the declaration of thrown exception '" + thrownClassName + "' which is a runtime exception.";
-    } else if (isDeclaredMoreThanOnce(thrownClass, thrownClasses)) {
-      return "Remove the redundant '" + thrownClassName + "' thrown exception declaration(s).";
+  private void processMethodCall(Symbol methodSymbol) {
+    if (methodSymbol.isMethodSymbol()) {
+      for (Type thrownType : ((Symbol.MethodSymbol) methodSymbol).thrownTypes()) {
+        registerThrownType(thrownType);
+      }
     }
-    return null;
   }
 
-  private static boolean isDeclaredMoreThanOnce(AsmClass thrownClass, List<AsmClass> thrownClassCandidates) {
-    int matches = 0;
+  @Override
+  public void leaveNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.CONSTRUCTOR, Tree.Kind.METHOD)) {
+      MethodTree methodTree = (MethodTree) tree;
+      List<TypeTree> exceptionsTree = new ArrayList<>(methodTree.throwsClauses());
+      checkRuntimeExceptions(methodTree, exceptionsTree);
+      checkRedundantExceptions(methodTree, exceptionsTree);
+      checkOtherExceptions(methodTree, exceptionsTree);
+    }
+    if (tree.is(Tree.Kind.CLASS, Tree.Kind.ENUM, Tree.Kind.INTERFACE, Tree.Kind.CONSTRUCTOR, Tree.Kind.METHOD)) {
+      thrownExceptions.pop();
+    }
+  }
 
-    for (AsmClass thrownClassCandidate : thrownClassCandidates) {
-      if (thrownClass.equals(thrownClassCandidate)) {
-        matches++;
+  private void checkRuntimeExceptions(Tree tree, List<TypeTree> exceptionsTree) {
+    for (int i = exceptionsTree.size() - 1; i >= 0; i--) {
+      TypeTree exceptionTree = exceptionsTree.get(i);
+      Type exceptionType = exceptionTree.symbolType();
+      if (exceptionType.isSubtypeOf("java.lang.RuntimeException")) {
+        addIssue(tree, "Remove the declaration of thrown exception '" + exceptionType.fullyQualifiedName() + "' which is a runtime exception.");
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "Remove the declaration of thrown exception '" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1192)